storage/sql.go

```diff
@@ -755,10 +762,100 @@ func (s *TxStorage) PersistModel(ctx context.Context, m interface{}) error {
 		}
 
 	}
-	if _, err := s.tx.ModelContext(ctx, m).
-		OnConflict("do nothing").
-		Insert(); err != nil {
-		return xerrors.Errorf("persisting model: %w", err)
+	if s.upsert {
+		conflict, upsert := GenerateUpsertStrings(m)
+		if _, err := s.tx.ModelContext(ctx, m).
+			OnConflict(conflict).
+			Set(upsert).
+			Insert(); err != nil {
+			return xerrors.Errorf("upserting model: %w", err)
+		}
+	} else {
+		if _, err := s.tx.ModelContext(ctx, m).
+			OnConflict("do nothing").
+			Insert(); err != nil {
+			return xerrors.Errorf("persisting model: %w", err)
+		}
 	}
 	return nil
 }
+
+var matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+var matchPkTag = regexp.MustCompile(`\bpk\b`)
+
+func ToSnakeCase(str string) string {
+	snake := matchFirstCap.ReplaceAllString(str, "${1}_${2}")
+	snake = matchAllCap.ReplaceAllString(snake, "${1}_${2}")
+	return strings.ToLower(snake)
+}
+
+// GenerateUpsertString accepts a visor model and returns two string containing SQL that may be used
+// to upsert the model. The first string is the conflict statement and the second is the insert.
+//
+// Example given the below model:
+//
+// type SomeModel struct {
+// 	Height    int64  `pg:",pk,notnull,use_zero"`
+// 	MinerID   string `pg:",pk,notnull"`
+// 	StateRoot string `pg:",pk,notnull"`
+// 	OwnerID  string `pg:",notnull"`
+// 	WorkerID string `pg:",notnull"`
+// }
+//
+// The strings returned are:
+// conflict string:
+//	"(cid, height, state_root) DO UPDATE"
+// update string:
+// 	"owner_id = EXCLUDED.owner_id, worker_id = EXCLUDED.worker_id"
+func GenerateUpsertStrings(model interface{}) (string, string) {
+	// thank you kind open source maintainer: https://github.com/fatih/structs
+	st := structs.New(model)
+	var cf []string
+	var ucf []string
+
+	// walk all exported fields in the model
+	for name := range st.Map() {
```

Comment:
I think it would be safer to use go-pg's orm to parse the pg struct tags. We don't need to keep in sync with any modifiers that are introduced. That assumes go-pg exports all the information we need here.

Reply:
I think you're right, I didn't explore go-pg very much wrt this. I'll take another look.